tests/MongoDB.Driver.Core.Tests/Core/Clusters/ClusterTests.cs

```diff
@@ -481,6 +481,89 @@ public void DescriptionChanged_should_be_raised_when_the_description_changes()
             _capturedEvents.Any().Should().BeFalse();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void SelectServer_should_call_custom_selector_if_there_are_eligible_servers([Values(true, false)] bool async)
+        {
+            int numberOfCustomServerSelectorCalls = 0;
+            var customServerSelector = new DelegateServerSelector((c, s) =>
+            {
+                numberOfCustomServerSelectorCalls++;
+                var highestPortServer = s.OrderByDescending(x => ((DnsEndPoint)x.EndPoint).Port).First();
+
+                return new[] { highestPortServer }; // select server with highest port
+            });
+
+            var settings = _settings.With(postServerSelector: customServerSelector);
+            var subject = new StubCluster(settings, _mockServerFactory.Object, _capturedEvents);
+
+            subject.Initialize();
+            subject.SetServerDescriptions(
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27017)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27018)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27019)),
+                ServerDescriptionHelper.Connected(subject.Description.ClusterId, new DnsEndPoint("localhost", 27020)));
+            _capturedEvents.Clear();
+
+            for (int i = 0; i < 3; i++)
+            {
+                var selectedServer = SelectServerAttempt(
+                    subject,
+                    new DelegateServerSelector((c, s) => s), // do not filter servers
+                    async);
+
+                ((DnsEndPoint)selectedServer.EndPoint).Port.Should().Be(27020);
```

Comment:
consider splitting this kine onto 2:

    var selectedServerPort = ((DnsEndPoint)selectedServer.EndPoint).Port;
    selectedServerPort.Should().Be(27020);